src/EventStore.Core.XUnit.Tests/Scavenge/Infrastructure/Scenario.cs

```diff
@@ -0,0 +1,596 @@
+
+			return await RunInternalAsync(
+				getExpectedKeptRecords,
+				getExpectedKeptIndexEntries);
+		}
+
+		private async Task<DbResult> RunInternalAsync(
+			Func<DbResult, LogRecord[][]> getExpectedKeptRecords,
+			Func<DbResult, LogRecord[][]> getExpectedKeptIndexEntries) {
+
+			if (string.IsNullOrEmpty(_dbPath))
+				throw new Exception("call WithDbPath");
+
+			// something is unhappy with memdb.. to do with writing the footer of the new inmem chunk
+			var memDb = false;
+			var dbConfig = TFChunkHelper.CreateDbConfig(_dbPath, 0, chunkSize: 1024 * 1024, memDb: memDb);
+			var dbResult = _getDb(dbConfig);
+			var keptRecords = getExpectedKeptRecords != null
+				? getExpectedKeptRecords(dbResult)
+				: null;
+
+			var keptIndexEntries = getExpectedKeptIndexEntries != null
+				? getExpectedKeptIndexEntries(dbResult)
+				: keptRecords;
+
+			dbResult.Db.Config.WriterCheckpoint.Flush();
+			dbResult.Db.Config.ChaserCheckpoint.Write(dbResult.Db.Config.WriterCheckpoint.Read());
+			dbResult.Db.Config.ChaserCheckpoint.Flush();
+			dbResult.Db.Config.ReplicationCheckpoint.Write(dbResult.Db.Config.WriterCheckpoint.Read());
+			dbResult.Db.Config.ReplicationCheckpoint.Flush();
+
+			var indexPath = Path.Combine(_dbPath, "index");
+			var readerPool = new ObjectPool<ITransactionFileReader>(
+				objectPoolName: "ReadIndex readers pool",
+				initialCount: ESConsts.PTableInitialReaderCount,
+				maxCount: ESConsts.PTableMaxReaderCount,
+				factory: () => new TFChunkReader(dbResult.Db, dbResult.Db.Config.WriterCheckpoint));
+
+			IHasher lowHasher = new XXHashUnsafe();
+			IHasher highHasher = new Murmur3AUnsafe();
+			ILongHasher<string> hasher = new CompositeHasher<string>(
+				new XXHashUnsafe(),
+				new Murmur3AUnsafe());
+
+			var humanHashers = true;
+			if (humanHashers) {
+				lowHasher = new ConstantHasher(0);
+				highHasher = new HumanReadableHasher32();
+				hasher = new CompositeHasher<string>(new ConstantHasher(0), new HumanReadableHasher32());
+			}
+
+			var tableIndex = new TableIndex(
+				directory: indexPath,
+				lowHasher: lowHasher,
+				highHasher: highHasher,
+				memTableFactory: () => new HashListMemTable(PTableVersions.IndexV4, maxSize: 200),
+				tfReaderFactory: () => new TFReaderLease(readerPool),
+				ptableVersion: PTableVersions.IndexV4,
+				maxAutoMergeIndexLevel: 5,
+				maxSizeForMemory: 1, // convert everything to ptables immediately
+				maxTablesPerLevel: 2,
+				inMem: memDb);
+
+			IReadIndex readIndex = new ReadIndex(
+				bus: new NoopPublisher(),
+				readerPool: readerPool,
+				tableIndex: tableIndex,
+				streamInfoCacheCapacity: 100,
+				additionalCommitChecks: true,
+				metastreamMaxCount: 1,
+				hashCollisionReadLimit: Opts.HashCollisionReadLimitDefault,
+				skipIndexScanOnReads: Opts.SkipIndexScanOnReadsDefault,
+				replicationCheckpoint: dbResult.Db.Config.ReplicationCheckpoint);
+
+			readIndex.Init(dbResult.Db.Config.WriterCheckpoint.Read());
+
```

Comment:
a call to tableIndex.WaitForBackgroundTasks() is required after readIndex.Init() if we want to wait for the index to stabilize / all PTables to be dumped

Reply:
oh good spot :+1: